network/benchmark/src/lib.rs

```diff
@@ -0,0 +1,605 @@
+    DataReply(BenchmarkDataReply),
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct BenchmarkDataSend {
+    pub request_counter: u64, // A monotonically increasing counter to verify responses
+    pub send_micros: i64, // micro seconds since some epoch at a moment just before this message is sent
+    pub data: Vec<u8>,    // A vector of bytes to send in the request; zero length in reply
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct BenchmarkDataReply {
+    pub request_counter: u64, // A monotonically increasing counter to verify responses
+    pub send_micros: i64, // micro seconds since some epoch at a moment just before this message is sent
+    pub your_send_micros: i64, // the send_micros from the previous message
+}
+
+/// Counter for pending network events to the monitoring service (server-side)
+pub static PENDING_BENCHMARK_NETWORK_EVENTS: Lazy<IntCounterVec> = Lazy::new(|| {
+    register_int_counter_vec!(
+        "aptos_benchmark_pending_network_events",
+        "Counters for pending network events for benchmarking",
+        &["state"]
+    )
+    .unwrap()
+});
+
+// Get messages from the network and quickly shuffle them to N threads of workers.
+async fn source_loop(
+    network_requests: NetworkServiceEvents<BenchmarkMessage>,
+    work: async_channel::Sender<(NetworkId, Event<BenchmarkMessage>)>,
+) {
+    let network_events: Vec<_> = network_requests
+        .into_network_and_events()
+        .into_iter()
+        .map(|(network_id, events)| events.map(move |event| (network_id, event)))
+        .collect();
+    let mut network_events = futures::stream::select_all(network_events).fuse();
+
+    loop {
+        match network_events.next().await {
+            None => {
+                // fused stream will never return more
+                work.close();
+                return;
+            },
+            Some(x) => match work.send(x).await {
+                Ok(_) => {},
+                Err(se) => {
+                    warn!("benchmark source_loop work send: {}", se);
+                },
+            },
+        };
+    }
+}
+
+async fn handle_direct(
+    network_client: &NetworkClient<BenchmarkMessage>,
+    network_id: NetworkId,
+    peer_id: AccountAddress,
+    msg_wrapper: BenchmarkMessage,
+    time_service: TimeService,
+    shared: Arc<RwLock<BenchmarkSharedState>>,
+) {
+    match msg_wrapper {
+        BenchmarkMessage::DataSend(send) => {
+            let reply = BenchmarkDataReply {
+                request_counter: send.request_counter,
+                send_micros: time_service.now_unix_time().as_micros() as i64,
+                your_send_micros: send.send_micros,
+            };
+            let result = network_client.send_to_peer(
+                BenchmarkMessage::DataReply(reply),
+                PeerNetworkId::new(network_id, peer_id),
+            );
+            if let Err(err) = result {
+                direct_messages("rerr");
```

Comment:
Looks like `rerr` is the metric label for the response error, right? If so, it might be worth making a constant for it. We tend to do this for services that use counters and labels, e.g., https://github.com/aptos-labs/aptos-core/blob/d62901d24ded80940c30b6dcff93d7e0dff852e6/mempool/src/counters.rs#L16

Reply:
"rerr" was premature compression for "reply_error" and has been expanded